mlflow/types/utils.py

```diff
@@ -496,6 +496,30 @@ def _infer_spark_type(x) -> DataType:
                 for f in x.fields
             ]
         )
+    elif isinstance(x, pyspark.sql.types.MapType):
+        if data is None or col_name is None:
+            raise Exception("Cannot infer schema for MapType without data and column name.")
+        # Map MapType to StructType
+        # Note that MapType assumes all values are of same type,
+        # if they're not then spark picks the first item's type
+        # and tries to convert rest to that type.
+        # e.g.
+        # >>> spark.createDataFrame([{"col": {"a": 1, "b": "b"}}]).show()
+        # +-------------------+
+        # |                col|
+        # +-------------------+
+        # |{a -> 1, b -> null}|
+        # +-------------------+
+        keys = data.selectExpr(f"map_keys({col_name}) as keys").first().keys
+        return Object(
+            properties=[
+                Property(
+                    name=k,
+                    dtype=_infer_spark_type(x.valueType),
```

Comment:
will this throw if we give it a nested object? e.g.
```
spark.createDataFrame([{"col": {"a": {"b": 0}}}])
```

Reply:
That's a good point! This case becomes complicated, as we need to fetch each field to check the data type. Maybe we should throw a message to ask users to provide the schema for such complex cases.

Reply:
it looks like it's possible to select the key value from the map:
<img width="416" alt="Screenshot 2023-11-02 at 3 27 28 PM" src="https://github.com/mlflow/mlflow/assets/148037680/9afdffc5-b194-44df-ab30-bc4b4d5c2b5f">

maybe we could do something like:
```python
if isinstance(x.valueType, pyspark.sql.types.MapType):
    nested_data = data.select(col(col_name).getField(k))
    dtype = _infer_spark_type(x.valueType, nested_data, f"{col_name}[{k}]")
else:
    dtype = _infer_spark_type(x.valueType)
```

it does look a little messy though and also i'm not sure if it works haha

Reply:
I think it should work.
I actually start to think whether we should support MapType 🤣  Because it only accept the same valuetypes, and if it can't safely convert all values to the same type, it erases the value to null. And for objects specifically, we should infer if each property is optional or not, which requires scanning over spark dataFrame to be accurate, but that's time-consuming as well. Or we just support simple Str->Scalar mapping?
@harupy WDYT?

Reply:
I guess it's not exactly the same as a Python object—that's the `StructType` right? Maybe the best solution here is to add a simple `MapType` to ColSpec that corresponds to this Spark type

Reply:
Python dictionary corresponds to Scala's Map type, then is inferred as MapType in spark directly. StructType is something people manually constructs, I think it's just when you want some structured data type then you define your own StructType.

Reply:
Ah I see, that makes sense. Plus I guess even if we add a different MapType, we won't really be able to infer whether a dict was supposed to be a StructType or a MapType anyway.

Throwing with a nicer error message sounds good to me!